### DISCUSSION TITLE
Variable replacement inside .npmrc not working as implied by exposeAllEnv

### DISCUSSION THREAD

**Discussion by WolfSoko** [renovatebot/renovate#18800] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovatebot/github-action@34.20.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

GITHUB.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

My `.npmrc` has an environment replacement like: 

```
//npm.fontawesome.com/:_authToken=${FONTAWESOME_TOKEN}
```

Several approached to get the environment variable into the docker container where tried, but non worked successfully.

- setting `customEnvVariables` via renovate.config.js and via `RENOVATE_CUSTOM_ENV_VARIABLES` in renovate.yml. 
  -  The values where available inside the renovate docker process, but for the npm env replacement, the variable was not available. 
- Hand over the Token by prefixing `FONTAWESOME_TOKEN` with `RENOVATE`.   `RENOVATE_FONTAWESOME_TOKEN`
  - worked when using RENOVATE_FONTAWESOME_TOKEN inside npmrc. 
  - But I can not change the TOKEN name for business reasons

As a workaround, it was possible to overwrite the `.npmrc` within the `renovate.json`. 
```
"npmrc": "@fortawesome:registry=https://npm.fontawesome.com/\n//npm.fontawesome.com/:_authToken=${RENOVATE_FONTAWESOME_TOKEN}",
```

This is no satisfiing workaround, because of strong coupling between `renovate.json` and `renovate.yml`. 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Setting npmrc (repository=*)
 WARN: Failed to replace env in config: ${FONTAWESOME_TOKEN} (repository=*)
...
ERROR: lookupUpdates error (repository=*, packageFile=package.json)
       "currentDigest": undefined,
       "currentValue": "14.2.9",
       "datasource": "npm",
       "depName": "@angular/animations",
       "digestOneAndOnly": undefined,
       "followTag": null,
       "lockedVersion": "14.2.9",
       "pinDigests": false,
       "rollbackPrs": true,
       "isVulnerabilityAlert": undefined,
       "updatePinnedDependencies": true,
       "unconstrainedValue": false,
       "err": {
         "message": "env-replace",
         "stack": "Error: env-replace\n    at /usr/src/app/node_modules/renovate/lib/modules/datasource/npm/npmrc.ts:30:13\n    at RE2.replace (<anonymous>)\n    at RE2.Symbol.replace.RE2.<computed> (/usr/src/app/node_modules/re2/re2.js:16:19)\n    at String.replace (<anonymous>)\n    at envReplace (/usr/src/app/node_modules/renovate/lib/modules/datasource/npm/npmrc.ts:27:16)\n    at setNpmrc (/usr/src/app/node_modules/renovate/lib/modules/datasource/npm/npmrc.ts:152:22)\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:271:15)\n    at getRawReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:332:23)\n    at getPkgReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:352:13)\n    at lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:80:52)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:58:34)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
       }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
the github-action is running the slim renovate inside a docker container, so not all host env is passed to the renovate container nor to the sidecar containers which renovate starts.

- https://github.com/renovatebot/renovate/discussions/18568#discussioncomment-3967900

**Reply by WolfSoko** [reply]
Thx for the hint. But shouldn't the customEnvVariables be available in the .npmrc replacement process?

**Reply by viceice** [reply]
no, those would be passed to child process, so npm could do it itself. but for renovate you should add a hostrule for query for updates

https://github.com/renovatebot/renovate/blob/d869c946d164965393f1259db5171e2a0303fb27/lib/modules/datasource/npm/npmrc.ts#L134-L159

**Reply by WolfSoko** [reply]
@viceice thank you very much for your fast replies. 

So I should disable the exposeAllEnv setting (therfore not getting the replacement error anymore) and try to use hostrules?

Looking at the code I think the automatic .npmrc extraction of hostRules would set the hostRule token value to '${FONTAWESOME_TOKEN}' in my case.

And the child process will have the customEnvVariables? Is this correct?

**Reply by WolfSoko** [reply]
Disabling exposeAllEnv in combination with hostRules worked. 
Thx for the input!
